include/envoy/upstream/upstream.h

```diff
@@ -318,6 +318,13 @@ class ClusterInfo {
    *         stats that will be freed when the cluster is removed.
    */
   virtual Stats::Scope& statsScope() const PURE;
+
+  /**
+   * Returns an optional source address for upstream connections to bind to.
+   *
+   * @return an optional source address to bind to.
+   */
+  virtual const Optional<Network::Address::InstanceConstSharedPtr>& sourceAddress() const PURE;
```

Comment:
Right now, HostImpl::createConnection is called with the ClusterInfo, dispatcher, and address.
I could change it to have ever caller of createConnection also pass in an optional source address and do all the plumbing that requires if you prefer.  This both seemed simpler and lent itself to the eventual per-cluster binding which was discussed on the issue.